open_spiel/python/pytorch/ppo.py

```diff
@@ -0,0 +1,344 @@
+    self.rewards = torch.zeros((self.steps_per_batch, self.num_envs)).to(device)
+    self.dones = torch.zeros((self.steps_per_batch, self.num_envs)).to(device)
+    self.values = torch.zeros((self.steps_per_batch, self.num_envs)).to(device)
+
+    # Initialize counters 
+    self.cur_batch_idx = 0
+    self.total_steps_done = 0
+    self.updates_done = 0
+    self.start_time = time.time()
+
+  def get_value(self, x):
+    return self.network.get_value(x)
+
+  def get_action_and_value(self, x, legal_actions_mask=None, action=None):
+    return self.network.get_action_and_value(x, legal_actions_mask, action)
+
+  def step(self, time_step, is_evaluation=False):
+    if is_evaluation:
+      singular_env = False
+      if not isinstance(time_step, list):
+        time_step = [time_step]
+        singular_env = True
+
+      with torch.no_grad():
+        legal_actions_mask = legal_actions_to_mask(
+          [ts.observations['legal_actions'][self.player_id] for ts in time_step], self.num_actions
+        ).to(self.device)
+        obs = torch.Tensor(np.array([ts.observations['info_state'][self.player_id] for ts in time_step])).to(self.device)
+        action, log_prob, entropy, value, probs = self.get_action_and_value(obs, legal_actions_mask=legal_actions_mask)
+
+        if singular_env:
+          return StepOutput(action=action[0].item(), probs=probs[0])
+        else:
+          return [StepOutput(action=a.item(), probs=p) for (a, p) in zip(action, probs)]
+    else:
+      with torch.no_grad():
+        # act
+        obs = torch.Tensor(np.array([ts.observations['info_state'][self.player_id] for ts in time_step])).to(self.device)
+        legal_actions_mask = legal_actions_to_mask(
+          [ts.observations['legal_actions'][self.player_id] for ts in time_step], self.num_actions
+        ).to(self.device)
+        action, logprob, _, value, probs = self.get_action_and_value(obs, legal_actions_mask=legal_actions_mask)
+
+        # store
+        self.legal_actions_mask[self.cur_batch_idx] = legal_actions_mask
+        self.obs[self.cur_batch_idx] = obs
+        self.actions[self.cur_batch_idx] = action
+        self.logprobs[self.cur_batch_idx] = logprob
+        self.values[self.cur_batch_idx] = value.flatten()
+
+        agent_output = [StepOutput(action=a.item(), probs=p) for (a, p) in zip(action, probs)]
+        return agent_output
+
+
+  def post_step(self, reward, done):
+    self.rewards[self.cur_batch_idx] = torch.tensor(reward).to(self.device).view(-1)
+    self.dones[self.cur_batch_idx] = torch.tensor(done).to(self.device).view(-1)
+
+    self.total_steps_done += self.num_envs
+    self.cur_batch_idx += 1
+
+      
+  def learn(self, time_step):
+    next_obs = torch.Tensor(np.array([ts.observations['info_state'][self.player_id] for ts in time_step])).to(self.device)
+
+    # Annealing the rate if instructed to do so.
+    if self.num_annealing_updates is not None:
+      frac = 1.0 - (self.updates_done) / self.num_annealing_updates
```

Comment:
Make sure LR doesn't go to 0 / negative ?

Reply:
It wouldn't go to 0 / negative when `num_annealing_updates=num_updates`:

```
learning_rate = 2.5e-4
total_timesteps = 500000
num_envs = 4
num_steps = 128
num_updates = total_timesteps // (num_envs * num_steps)
num_annealing_updates = num_updates
print(f"num_updates={num_updates}")
for update in range(num_updates):
    # Annealing the rate if instructed to do so.
    frac = 1.0 - (update) / num_annealing_updates
    lrnow = frac * learning_rate
print(f"frac={frac}")
frac = 1.0 - (update+1) / num_annealing_updates
print(f"do an extra annealing step is not necessary because frac would become {frac}")
```
```
num_updates=976
frac=0.0010245901639344135
do an extra annealing step is not necessary because frac would become 0.0
```

However, if `num_annealing_updates` is set to `num_updates - 2`, then 0 / negative happens. Maybe instead of a `ValueError` we can check the `num_annealing_updates` is greater than `num_updates - 1` @newmanne?

```
num_updates=976
frac=-0.0010266940451746365
do an extra annealing step is not necessary because frac would become -0.002053388090349051
```

Reply:
We've moved the learning rate annealing outside of the agent. By default there is no annealing, and it can be put into the training loop if the user wants (as seen in `ppo_example.py` with the `anneal_lr` flag.